ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java

```diff
@@ -749,7 +839,140 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
+   *                              {@link ByteBuf}.
+   * @throws Exception
+   */
+  private void backPressureTest(String digestAlgorithm, boolean useCopyForcingByteBuf)
+      throws Exception {
+    List<HttpContent> httpContents = new ArrayList<HttpContent>();
+    byte[] contentBytes = RestTestUtils.getRandomBytes(GENERATED_CONTENT_SIZE);
+    ByteBuffer content = ByteBuffer.wrap(contentBytes);
+    splitContent(contentBytes, GENERATED_CONTENT_PART_COUNT, httpContents, useCopyForcingByteBuf);
+    int chunkSize = httpContents.get(0).content().readableBytes();
+    int[] bufferWatermarks = {1, chunkSize - 1, chunkSize,
+        chunkSize + 1, chunkSize * httpContents.size() / 2, content.limit() - 1, content.limit(), content.limit() + 1};
+    for (int bufferWatermark : bufferWatermarks) {
+      NettyRequest.bufferWatermark = bufferWatermark;
+      // start reading before addition of content
+      httpContents.clear();
+      content.rewind();
+      splitContent(contentBytes, GENERATED_CONTENT_PART_COUNT, httpContents, useCopyForcingByteBuf);
+      doBackPressureTest(digestAlgorithm, content, httpContents, 0);
+      // start reading in the middle of content add
+      httpContents.clear();
+      content.rewind();
+      splitContent(contentBytes, GENERATED_CONTENT_PART_COUNT, httpContents, useCopyForcingByteBuf);
+      doBackPressureTest(digestAlgorithm, content, httpContents, httpContents.size() / 2);
+      // start reading after all content added
+      httpContents.clear();
+      content.rewind();
+      splitContent(contentBytes, GENERATED_CONTENT_PART_COUNT, httpContents, useCopyForcingByteBuf);
+      doBackPressureTest(digestAlgorithm, content, httpContents, httpContents.size());
+    }
+  }
+
+  /**
+   * Does the backpressure test by ensuring that {@link Channel#read()} isn't called when the number of bytes buffered
+   * is above the {@link NettyRequest#bufferWatermark}. Also ensures that {@link Channel#read()} is called correctly
+   * when the number of buffered bytes falls below the {@link NettyRequest#bufferWatermark}.
+   * @param digestAlgorithm the digest algorithm to use. Can be empty or {@code null} if digest checking is not
+   *                        required.
+   * @param content the complete content.
+   * @param httpContents {@code content} in parts and as {@link HttpContent}. Should contain all the data in
+   * {@code content}.
+   * @param numChunksToAddBeforeRead the number of {@link HttpContent} to add before making the
+   * {@link NettyRequest#readInto(AsyncWritableChannel, Callback)} call.
+   * @throws Exception
+   */
+  private void doBackPressureTest(String digestAlgorithm, ByteBuffer content, List<HttpContent> httpContents,
+      int numChunksToAddBeforeRead)
+      throws Exception {
+    if (numChunksToAddBeforeRead < 0 || numChunksToAddBeforeRead > httpContents.size()) {
+      throw new IllegalArgumentException("Illegal value of numChunksToAddBeforeRead");
```

Comment:
This function is meant to be used by multiple tests and multiple cases - so i thought it is useful to have the check here to avoid spurious failures.